# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/x64/MacroAssembler-x64.cpp
# Commit: eeaeb40f8b86
# Full Hash: eeaeb40f8b869121eff65b1c2d11af423c9a9ecc
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2022-03-09 09:44:44
# Regressor Bug: 885514
# File Overlap Count: 1
# Description:
#   Bug 885514: Simplify JSOp::Retsub r=jandem
#   
#   Currently, JSOp::Retsub takes two arguments, and uses one to decide whether to throw the second or use it as a resume index. This patch splits out the exception-handling logic into explicit bytecode, leaving JSOp::Retsub as a simple indirect jump.
#   
#   To make the generated bytecode work out nicely, I've swapped the order of the arguments: now `throwing` is on top, and the exception/resumeindex argument is underneath it.
# ==============================================================================

diff -r 1cb0b352b024 -r eeaeb40f8b86 js/src/jit/x64/MacroAssembler-x64.cpp
--- a/js/src/jit/x64/MacroAssembler-x64.cpp	Tue Mar 08 21:10:22 2022 +0000
+++ b/js/src/jit/x64/MacroAssembler-x64.cpp	Tue Mar 08 21:10:22 2022 +0000
@@ -526,9 +526,8 @@
   loadPtr(Address(rsp, offsetof(ResumeFromException, stackPointer)), rsp);
   jmp(Operand(rax));
 
-  // If we found a finally block, this must be a baseline frame. Push
-  // two values expected by JSOp::Retsub: BooleanValue(true) and the
-  // exception.
+  // If we found a finally block, this must be a baseline frame. Push two
+  // values expected by JSOp::Retsub: the exception and BooleanValue(true).
   bind(&finally);
   ValueOperand exception = ValueOperand(rcx);
   loadValue(Address(esp, offsetof(ResumeFromException, exception)), exception);
@@ -537,8 +536,8 @@
   loadPtr(Address(rsp, offsetof(ResumeFromException, framePointer)), rbp);
   loadPtr(Address(rsp, offsetof(ResumeFromException, stackPointer)), rsp);
 
+  pushValue(exception);
   pushValue(BooleanValue(true));
-  pushValue(exception);
   jmp(Operand(rax));
 
   // Only used in debug mode. Return BaselineFrame->returnValue() to the caller.